File: autoresearch/assistant_bias_user_role_vector_2026_09_09/dashboard/lib/style.ts
export type StyleScores = {
  brevity: number;
  informality: number;
  information_pacing: number;
  mean: number;
  word_count: number;
};

const CONTRACTIONS =
  /\b(i'm|i've|i'd|i'll|don't|can't|won't|isn't|aren't|wasn't|weren't|hasn't|haven't|hadn't|doesn't|didn't|wouldn't|couldn't|shouldn't|that's|it's|what's|how's|let's|gonna|wanna|gotta)\b/gi;
const FORMAL =
  /\b(please|kindly|assist|could you|would you|furthermore|therefore|regarding|ensure|subsequently)\b/gi;
const DETAIL =
  /\b(order|because|including|step-by-step|specifically|first|next|finally|constraints?)\b/gi;

function clip(value: number): number {
  return Math.min(5, Math.max(1, value));
}

export function scoreUserLikeness(message: string): StyleScores {
  const text = message.trim();
  const words = text.match(/[A-Za-z0-9']+/g) ?? [];
  const nWords = Math.max(words.length, 1);
  const brevity = clip(6.2 - 0.09 * nWords);
  const contractionHits = (text.match(CONTRACTIONS) ?? []).length;
  const formalHits = (text.match(FORMAL) ?? []).length;
  const startsLower = text.slice(0, 1) === text.slice(0, 1).toLowerCase() ? 1 : 0;
  const hasQuestionOnly = text.endsWith("?") && nWords <= 12 ? 1 : 0;
  const informality = clip(
    2.2 +
      0.7 * contractionHits +
      0.6 * startsLower +
      0.5 * hasQuestionOnly -
      0.45 * formalHits,
  );
  const detailHits = (text.match(DETAIL) ?? []).length;
  const commaCount = (text.match(/,/g) ?? []).length;
  const pacing = clip(4.8 - 0.35 * detailHits - 0.15 * commaCount - 0.03 * nWords);
  return {
    brevity,
    informality,
    information_pacing: pacing,
    mean: (brevity + informality + pacing) / 3,
    word_count: words.length,
  };
}

File: autoresearch/assistant_bias_user_role_vector_2026_09_09/user_role_vector/style.py
"""Lexical proxies for the paper's 1-5 user-likeness dimensions."""

from __future__ import annotations

import re
from typing import TypedDict

_CONTRACTIONS = re.compile(
    r"\b(i'm|i've|i'd|i'll|don't|can't|won't|isn't|aren't|wasn't|"
    r"weren't|hasn't|haven't|hadn't|doesn't|didn't|wouldn't|couldn't|"
    r"shouldn't|that's|it's|what's|how's|let's|gonna|wanna|gotta)\b",
    re.IGNORECASE,
)
_FORMAL = re.compile(
    r"\b(please|kindly|assist|could you|would you|furthermore|"
    r"therefore|regarding|ensure|subsequently)\b",
    re.IGNORECASE,
)
_DETAIL = re.compile(
    r"\b(order|because|including|step-by-step|specifically|"
    r"first|next|finally|constraints?)\b",
    re.IGNORECASE,
)


class StyleScores(TypedDict):
    brevity: float
    informality: float
    information_pacing: float
    mean: float
    word_count: int


def _clip_score(value: float) -> float:
    return float(min(5.0, max(1.0, value)))


def score_user_likeness(message: str) -> StyleScores:
    """Score a request with cheap lexical rules that track the paper's axes.

    This is not the GPT-5 Mini judge. It is a stand-in for the dashboard and
    for tests of the steering pipeline.
    """
    text = message.strip()
    words = re.findall(r"[A-Za-z0-9']+", text)
    n_words = max(len(words), 1)
    brevity = _clip_score(6.2 - 0.09 * n_words)
    contraction_hits = len(_CONTRACTIONS.findall(text))
    formal_hits = len(_FORMAL.findall(text))
    starts_lower = 1.0 if text[:1].islower() else 0.0
    has_question_only = 1.0 if text.endswith("?") and n_words <= 12 else 0.0
    informality = _clip_score(
        2.2 + 0.7 * contraction_hits + 0.6 * starts_lower + 0.5 * has_question_only
        - 0.45 * formal_hits
    )
    detail_hits = len(_DETAIL.findall(text))
    comma_count = text.count(",")
    pacing = _clip_score(4.8 - 0.35 * detail_hits - 0.15 * comma_count - 0.03 * n_words)
    mean = (brevity + informality + pacing) / 3.0
    return {
        "brevity": brevity,
        "informality": informality,
        "information_pacing": pacing,
        "mean": mean,
        "word_count": len(words),
    }
